compressors/baseline_compressors.py

```diff
@@ -3,71 +3,195 @@
+
+    def __init__(self):
+        super().__init__()
+        self.alphabet_encoder = AlphabetEncoder()
+
+    def encode_block(self, data_block: DataBlock):
+        """first encode the alphabet and then each data symbol using fixed number of bits"""
+        # get bit width
+        alphabet = data_block.get_alphabet()
+
+        # encode alphabet
+        encoded_bitarray = self.alphabet_encoder.encode_block(alphabet)
+
+        # encode data
+        symbol_bit_width = get_bit_width(len(alphabet))
+        alphabet_dict = {a: i for i, a in enumerate(alphabet)}
+        for s in data_block.data_list:
+            encoded_bitarray += uint_to_bitarray(alphabet_dict[s], bit_width=symbol_bit_width)
+
+        return encoded_bitarray
+
+
+class FixedBitwidthDecoder(DataDecoder):
+    def __init__(self):
+        super().__init__()
+        self.alphabet_decoder = AlphabetDecoder()
+
+    def decode_block(self, bitarray: BitArray):
+        """Decode data encoded by FixedBitwidthDecoder
+
+        - retrieve the alphabet
+        - the size of the alphabet implies the bit width used for encoding the symbols
+        - decode data
+        """
+        # get the alphabet
+        alphabet, num_bits_consumed = self.alphabet_decoder.decode_block(bitarray)
+
+        # decode data
+        symbol_bit_width = get_bit_width(len(alphabet))
+
+        data_list = []
+        while num_bits_consumed < len(bitarray):
+            symbol_bitarray = bitarray[num_bits_consumed : (num_bits_consumed + symbol_bit_width)]
+            ind = bitarray_to_uint(symbol_bitarray)
+            data_list.append(alphabet[ind])
+            num_bits_consumed += symbol_bit_width
+
+        return DataBlock(data_list), num_bits_consumed
+
+
+#########################################
+
+
+def test_alphabet_encode_decode():
+    """test the alphabet compression"""
+    # define encoder, decoder
+    encoder = AlphabetEncoder()
+    decoder = AlphabetDecoder()
+
+    # create some sample data
+    alphabet = ["A", "B", "C"]
+    output_bits_block = encoder.encode_block(alphabet)
+    decoded_alphabet, num_bits_consumed = decoder.decode_block(output_bits_block)
+    assert alphabet == decoded_alphabet
+    assert num_bits_consumed == (1 + len(alphabet)) * 8
+
+
+def test_fixed_bitwidth_encode_decode():
+    """test the encode_block and decode_block functions of FixedBitWidthEncoder and FixedBitWidthDecoder"""
+    # define encoder, decoder
+    encoder = FixedBitwidthEncoder()
+    decoder = FixedBitwidthDecoder()
+
+    # create some sample data
+    data_list = ["A", "B", "C", "C", "A", "C"]
+    data_block = DataBlock(data_list)
+
+    is_lossless, codelen = try_lossless_compression(data_block, encoder, decoder)
+    assert is_lossless
+
+    # check if the length of the encoding was correct
+    alphabet_bits = (1 + len(data_block.get_alphabet())) * 8
+    assert codelen == len(data_list) * 2 + alphabet_bits
+
+
+def test_fixed_bitwidth_file_write():
```

Comment:
Should we move this to test_utils (similar to try_lossless_compression) since this is useful even beyond the fixed bitwidth compressor? And this is also how we would typically use the compressor, so having it outside the bitwidth file can make sense. 
One question I wonder about is whether encode_block/decode_block passing roundtrip implies that encode/decode roundtrip always passes. I think that's true but would like to confirm with you once.

Reply:
Yes, I will move this to utils.
In general encode_block/decode_block roundtrip would imply encode/decode always passes. We however need to make sure the "DataStream" we are using is able to read/write correctly. So probably need a test for the DataStream as well. 

Also need to think how to abstract this correctly for different datastreams ... (we will most commonly have TextDataStream, but we could have others.. so need to think about that)

Reply:
Perhaps some way to pass the stream type to the test function?

Reply:
I refactored this part of the code. @shubhamchandak94 @tpulkit 
- Added utility function `encode_file` and `decode_file` to DataEncoder and DataDecoder. This is a bit of a change to the API, so do take a look and let me know if this makes sense. I am not 100% sure about this, so would be good to take a look and see if there is a way to improve.
- Added utility functions to `test_utils.py`